packages/visibility_detector/lib/src/render_visibility_detector.dart

```diff
@@ -4,28 +4,92 @@
 // license that can be found in the LICENSE file or at
 // https://developers.google.com/open-source/licenses/bsd
 
+import 'dart:async';
+import 'dart:math' as math;
+
 import 'package:flutter/foundation.dart';
 import 'package:flutter/rendering.dart';
+import 'package:flutter/scheduler.dart';
 
 import 'visibility_detector.dart';
-import 'visibility_detector_layer.dart';
+import 'visibility_detector_controller.dart';
 
-/// The [RenderObject] corresponding to the [VisibilityDetector] widget.
-///
-/// [RenderVisibilityDetector] is a bridge between [VisibilityDetector] and
-/// [VisibilityDetectorLayer].
-class RenderVisibilityDetector extends RenderProxyBox {
-  /// Constructor.  See the corresponding properties for parameter details.
-  RenderVisibilityDetector({
-    RenderBox? child,
-    required this.key,
-    required VisibilityChangedCallback? onVisibilityChanged,
-  })  : assert(key != null),
-        _onVisibilityChanged = onVisibilityChanged,
-        super(child);
+mixin RenderVisibilityDetectorBase on RenderObject {
+  static int? get debugUpdateCount {
+    if (!kDebugMode) {
+      return null;
+    }
+    return _updates.length;
+  }
+
+  static Map<Key, VoidCallback> _updates = <Key, VoidCallback>{};
+  static Map<Key, VisibilityInfo> _lastVisibility = <Key, VisibilityInfo>{};
+
+  /// See [VisibilityDetectorController.notifyNow].
+  static void notifyNow() {
+    _timer?.cancel();
+    _timer = null;
+    _processCallbacks();
+  }
+
+  static void forget(Key key) {
+    _updates.remove(key);
+
+    if (_updates.isEmpty) {
+      _timer?.cancel();
+      _timer = null;
+    }
+  }
+
+  static Timer? _timer;
+  static void _handleTimer() {
+    _timer = null;
+    // Ensure that work is done between frames so that calculations are
+    // performed from a consistent state.  We use `scheduleTask<T>` here instead
+    // of `addPostFrameCallback` or `scheduleFrameCallback` so that work will
+    // be done even if a new frame isn't scheduled and without unnecessarily
+    // scheduling a new frame.
+    SchedulerBinding.instance.scheduleTask<void>(
+      _processCallbacks,
+      Priority.touch,
+    );
+  }
+
+  /// Executes visibility callbacks for all updated instances.
+  static void _processCallbacks() {
+    for (final callback in _updates.values) {
+      callback();
+    }
+    _updates.clear();
+  }
+
+  void _fireCallback(ContainerLayer? layer, Rect bounds) {
+    final oldInfo = _lastVisibility[key];
+    final info = _determineVisibility(layer, bounds);
+    final visible = !info.visibleBounds.isEmpty;
+
+    if (oldInfo == null) {
+      if (!visible) {
+        return;
+      }
+    } else if (info.matchesVisibility(oldInfo)) {
+      return;
+    }
+
+    if (visible) {
+      _lastVisibility[key] = info;
+    } else {
+      // Track only visible items so that the map does not grow unbounded.
+      _lastVisibility.remove(key);
+    }
+
+    onVisibilityChanged?.call(info);
```

Comment:
The VisibilityInfo object contains the newly added `screenRect` value. It appears that there is a code pass where the `screenRect` would change, but the `onVisibilityChanged` is not called (because of the early return above when matchesVisibility is true) to inform the owner about the callback about it, leaving them with potential outdated information. This could be surprising. 

Not sure what the motivation was to add screenRect to that data class, but having an outdated value there could be a source of future bugs.

Reply:
Ahhh, I wanted to get rid of a global static dictionary. 

The only thing that uses these are unit tests in this library. I suppose I could just drop that entirely. Otherwise we'd have to emit changes when just the screen rect changes, which maybe is fine but maybe is a small change in behavior that will cause problems for google consumers.

Reply:
So a case where I think you're right: if the widget resizes, but the resizing happens outside of the current clip. The visible bounds won't change but the widget bounds would.

I'm going to remove this as it's only used in tests.

Reply:
This remains my main concern as I can already see how people will rely on this and then get surprised if it doesn't update properly. I'd either remove this feature alltogether, or - if it is only used in test - make it a debug only feature that throws in release when accessed.